crates/turborepo-filewatch/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+use notify::{RecursiveMode, Watcher};
+use notify_debouncer_full::{DebounceEventResult, DebouncedEvent, Debouncer, FileIdMap};
+use thiserror::Error;
+use tokio::sync::{broadcast, mpsc};
+// windows -> no recursive watch, watch ancestors
+// linux -> recursive watch, watch ancestors
+#[cfg(feature = "watch_ancestors")]
+use turbopath::PathRelation;
+use turbopath::{AbsoluteSystemPath, AbsoluteSystemPathBuf};
+// macos -> custom watcher impl in fsevents, no recursive watch, no watching ancestors
+#[cfg(target_os = "macos")]
+use {
+    fsevent::FsEventWatcher,
+    notify_debouncer_full::{new_debouncer_opt, DebounceEventHandler},
+};
+#[cfg(feature = "watch_recursively")]
+use {
+    notify::event::{CreateKind, Event, EventAttributes},
+    std::{path::Path, time::Instant},
+    walkdir::WalkDir,
+};
+#[cfg(not(target_os = "macos"))]
+use {notify::RecommendedWatcher, notify_debouncer_full::new_debouncer};
+
+#[cfg(target_os = "macos")]
+mod fsevent;
+
+#[cfg(not(target_os = "macos"))]
+type Backend = RecommendedWatcher;
+#[cfg(target_os = "macos")]
+type Backend = FsEventWatcher;
+
+#[derive(Debug, Error)]
+pub enum WatchError {
+    #[error("filewatching backend error: {0}")]
+    Notify(#[from] notify::Error),
+    #[error("filewatching stopped")]
+    Stopped(#[from] std::sync::mpsc::RecvError),
+    #[error("enumerating recursive watch: {0}")]
+    WalkDir(#[from] walkdir::Error),
+    #[error("filewatching failed to start: {0}")]
+    Setup(String),
+}
+
+// We want to broadcast the errors we get, but notify::Error does not implement
+// Clone. We provide a wrapper that uses an Arc to implement Clone so that we
+// can send errors on a broadcast channel.
+#[derive(Clone, Debug)]
+pub struct NotifyError(Arc<notify::Error>);
+
+impl From<notify::Error> for NotifyError {
+    fn from(value: notify::Error) -> Self {
+        Self(Arc::new(value))
+    }
+}
+
+pub struct FileSystemWatcher {
+    sender: broadcast::Sender<Result<DebouncedEvent, Vec<NotifyError>>>,
+    // _exit_ch exists to trigger a close on the receiver when an instance
+    // of this struct is dropped. The task that is receiving events will exit,
+    // dropping the other sender for the broadcast channel, causing all receivers
+    // to be notified of a close.
+    _exit_ch: tokio::sync::oneshot::Sender<()>,
+}
+
+impl FileSystemWatcher {
+    pub fn new(root: &AbsoluteSystemPath) -> Result<Self, WatchError> {
+        let (sender, _) = broadcast::channel(1024);
+        let (send_file_events, mut recv_file_events) = mpsc::channel(1024);
+        let watch_root = root.to_owned();
+        let broadcast_sender = sender.clone();
+        let debouncer = run_watcher(&watch_root, send_file_events).unwrap();
+        let (exit_ch, exit_signal) = tokio::sync::oneshot::channel();
+        // Ensure we are ready to receive new events, not events for existing state
+        futures::executor::block_on(wait_for_cookie(&watch_root, &mut recv_file_events))?;
```

Comment:
I expect *this component* to block on a filesystem cookie. TBD on what the daemon will do, but it will be up to the daemon code to block or stick this in a future.

Reply:
I'm fine with kicking this down the road. We'll just need to be mindful that throwing this call in a future will starve out any other futures that are scheduled in that thread until this call finishes.